cluster-autoscaler/core/static_autoscaler_csi_test.go

```diff
@@ -0,0 +1,433 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	sdplanner "k8s.io/autoscaler/cluster-autoscaler/core/scaledown/planner"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/clustersnapshot/predicate"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/clustersnapshot/store"
+	csinodeprovider "k8s.io/autoscaler/cluster-autoscaler/simulator/csi/provider"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/drainability/rules"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/framework"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/options"
+	. "k8s.io/autoscaler/cluster-autoscaler/utils/test"
+	informers "k8s.io/client-go/informers"
+	clientsetfake "k8s.io/client-go/kubernetes/fake"
+	v1storagelister "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/utils/ptr"
+)
+
+type testNodeGroupCSI struct {
+	name                string
+	cpu, mem            int64
+	csiNodeTemplateFunc func(nodeName string) *storagev1.CSINode
+}
+
+type fakeCSINodeLister struct {
+	nodes map[string]*storagev1.CSINode
+}
+
+var _ v1storagelister.CSINodeLister = (*fakeCSINodeLister)(nil)
+
+func (l *fakeCSINodeLister) List(_ labels.Selector) ([]*storagev1.CSINode, error) {
+	out := make([]*storagev1.CSINode, 0, len(l.nodes))
+	for _, n := range l.nodes {
+		out = append(out, n)
+	}
+	return out, nil
+}
+
+func (l *fakeCSINodeLister) Get(name string) (*storagev1.CSINode, error) {
+	n, ok := l.nodes[name]
+	if !ok {
+		return nil, fmt.Errorf("csinode %q not found", name)
+	}
+	return n, nil
+}
+
+func TestStaticAutoscalerCSI(t *testing.T) {
+	now := time.Now()
+
+	const (
+		driverName     = "ebs.csi.aws.com"
+		storageClass   = "sc-ebs"
+		volumeLimit    = int32(10)
+		templateVolUse = 2
+		podVolUse      = 6
+	)
+
+	node1GroupCSI := &testNodeGroupCSI{
+		name:                "node1CSINode",
+		cpu:                 1000,
+		mem:                 1000,
+		csiNodeTemplateFunc: csiNodeTemplate(driverName, ptr.To(volumeLimit)),
+	}
+
+	testCases := map[string]struct {
+		nodeGroups         map[*testNodeGroupCSI]int
+		templatePods       map[string][]*apiv1.Pod
+		pods               []*apiv1.Pod
+		expectedScaleUps   map[string]int
+		expectedScaleDowns map[string][]string
+	}{
+		"scale-up: CSI volume limit blocks packing, requires 2 extra nodes": {
+			nodeGroups: map[*testNodeGroupCSI]int{node1GroupCSI: 1},
+			templatePods: map[string][]*apiv1.Pod{
+				// A DaemonSet-like pod that consumes some CSI volumes on every node.
+				node1GroupCSI.name: {buildCSITestPod("template-ds", templateVolUse, node1GroupCSI.name+"-0")},
+			},
+			// Two pending pods, each using podVolUse volumes.
+			pods: []*apiv1.Pod{
+				buildCSITestPod("pending-0", podVolUse, ""),
+				buildCSITestPod("pending-1", podVolUse, ""),
+				buildCSITestPod("pending-2", podVolUse, ""),
+			},
+			// One existing node has the CSI pod already; only one pending pod can fit because (5 + 3 + 3) > 10.
```

Comment:
Is the comment based on old values? The pending pods seem to consume 6 volumes each, and the template pod 2 volumes - so I'd expect `(6 + 6 + 2) > 10`